### PR TITLE
Fix batching issues

### LINKED ISSUE #4: Batch issue with valid spawns mask
It appears that valid_spawns_mask property is incorrect when batched. If valid_spawns_mask has a batch dimension, then
https://github.com/RoboEden/jux/blob/main/jux/map/board.py#L90-L96

is not right. It will set first and last spawn mask to 0, and do incorrect roll

### PR DESCRIPTION
**Pull request by StoneT2000**
Should fix #4 

I think @property isn't handled properly outside of jit. When user accesses a @property information e.g. valid_spawns_mask it won't work when the batched data has been returned outside of jit.

### CODE REVIEW

File: jux/map_generator/generator.py [hunk 1]
```diff
@@ -24,10 +24,14 @@ class GameMap(NamedTuple):
 
     @property
     def width(self) -> int:
+        if len(self.rubble.shape) == 3:
+            return self.rubble.shape[2]
         return self.rubble.shape[1]
```
Comment (liqimai) [inline jux/map_generator/generator.py:29]: I think a more general way is to access dimensions by negative index, because the batch dimensions are always leading dimension.
```python
    @property
    def width(self) -> int:
        return self.rubble.shape[-1]

    @property
    def height(self) -> int:
        return self.rubble.shape[-2]
```
There may be more than one batch dimension.

File: jux/map/board.py [hunk 2]
```diff
@@ -88,18 +88,27 @@ def ore(self) -> Array:
     @property
     def valid_spawns_mask(self) -> Array:  # bool[height, width]
         valid_spawns_mask = (~self.map.ice & ~self.map.ore)  # bool[height, width]
-        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, 1, axis=0)
-        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, -1, axis=0)
+        batched = True
```
Comment (liqimai) [inline jux/map/board.py:91]: Same as below. There may be more than one batch dimension, but batch dimensions are always leading dimension, so we can always access the dimensions we want by negative index. Here is an example of how to tackle more general cases without special treatment for some cases.
```python
    @property
    def valid_spawns_mask(self) -> Array:  # bool[height, width]
        valid_spawns_mask = (~self.map.ice & ~self.map.ore)  # bool[..., height, width]
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, 1, axis=-2)
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, -1, axis=-2)
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, 1, axis=-1)
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, -1, axis=-1)
        valid_spawns_mask = valid_spawns_mask.at[..., [0, -1], :].set(False)
        valid_spawns_mask = valid_spawns_mask.at[..., :, [0, -1]].set(False)

        batch_shape = factory_overlap.shape[:-3]
        factory_overlap = self.factory_pos[..., None, :] + delta_xy  # int[..., 2 * MAX_N_FACTORIES, 85, 2]
        factory_overlap = factory_overlap.reshape(batch_shape + (-1, 2))
        factory_overlap = jnp.clip(factory_overlap, 0, jnp.array([self.height - 1, self.width - 1]))
        valid_spawns_mask = valid_spawns_mask.at[factory_overlap[..., 0], factory_overlap[..., 1]].set(False, mode='drop')
```

**Review by liqimai — CHANGES_REQUESTED** [review]

File: jux/map/board.py (hunk 2 above)
Comment (StoneT2000) [reply jux/map/board.py:91]: this still has a bug with factory_overlap[...] and setting those points to False on valid_spawns_mask.

In batch, valid_spawns_mask is shape `(b, 48, 48)`
factory_overlap is shape `(b, 1020, 2)`

`valid_spawns_mask.at[factory_overlap[..., 0], factory_overlap[..., 1]]` will incorrectly select the wrong indices. `factory_overlap[..., 0]` will select the batch index, and `factory_overlap[..., 1]` will select the first map dimension.

**Comment by StoneT2000** [comment]
added another bug fix for the `to_lux` function

**Review by liqimai — APPROVED** [review]
